boot/bootutil/src/bootutil_public.c

```diff
@@ -524,26 +523,24 @@ boot_set_pending_multi(int image_index, int permanent)
             rc = boot_write_swap_info(fap, swap_type, 0);
         }
 
-        flash_area_close(fap);
-        return rc;
+        break;
 
     case BOOT_MAGIC_BAD:
         /* The image slot is corrupt.  There is no way to recover, so erase the
          * slot to allow future upgrades.
          */
-        rc = flash_area_open(FLASH_AREA_IMAGE_SECONDARY(image_index), &fap);
-        if (rc != 0) {
-            return BOOT_EFLASH;
-        }
-
         flash_area_erase(fap, 0, fap->fa_size);
-        flash_area_close(fap);
-        return BOOT_EBADIMAGE;
+        rc = BOOT_EBADIMAGE;
+        break;
 
     default:
         assert(0);
-        return BOOT_EBADIMAGE;
+        rc = BOOT_EBADIMAGE;
     }
+
+done:
+    flash_area_close(fap);
+    return 0;
```

Comment:
Should probably return `rc`, no?

Reply:
(after lots of swearing to himself, he has finally admitted) Yes, you are right @utzig.